CloudMonitoring: Move data manipulation to backend

iwysiu opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Moves the data manipulation for the resource calls to the backend.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #41377 

**Special notes for your reviewer**:
- I added the `github.com/andybalholm/brotli` dependency for `br` reading/writing. It's the most used pure go library for this, and it's MIT licensed. However, I'm not sure where the license should be added in the Grafana repo.
- The go-cache library I added was already being used in repo.

iwysiu committed:
CloudMonitoring: Move data manipulation to backend

andresmgot reviewed (commented):
Great job here! One thing that I am unsure of is the use of the cache. That can lead to unexpected bugs. For example, if a user gets a forbidden error and adds the required permissions, they will still get the same error until the cache expires. Also, as far as I can see, the cache is only based on the URL of the request, what if now (or in the future) a resource requires something in the request body? cc/ @sunker in case I am missing some context here.

iwysiu committed:
comment regex and remove old test

iwysiu commented:
Added a comment to the regex!

I rebased to try to get rid of some local compilation errors, and the frontend test I removed was testing logic that's now in the backend.

marefr commented:
Guess I have some questions about the cache as well. Why do you need a cache in the first place?

iwysiu commented:
The cache is honestly because the original code was caching in the frontend (in api.ts), so I assumed there was significant enough traffic to warrant it. I'm ok with removing it.

andresmgot commented:
> The cache is honestly because the original code was caching in the frontend (in api.ts), so I assumed there was significant enough traffic to warrant it. I'm ok with removing it.

Got it, one difference is that before (in api.ts), only successful requests were being cached, now we would be caching all. I would still avoid caching to avoid issues but there may be some cost-related issue I am not aware of. @sunker can you jump in here? You have more context. Thanks!

sunker reviewed (commented):
Great work @iwysiu!! So happy to see the datasource.ts file cleaned up, and nice that you've moved the test datasource check to the backend too. 

>Guess I have some questions about the cache as well. Why do you need a cache in the first place?

Good question. Think it might be legacy from when this data source used to be in angular. 

Looking at the network traffic, it seems that projects are loaded three times when the query editor is rendered. I believe user pay per call to these api:s, so the cache does indeed have a purpose. However, the right way to solve this would be to prevent additional calls to get projects from within the react components. I think changing the Project.tsx file to something like this should do the trick:

```typescript
const [projects, setProjects] = useState<Array<SelectableValue<string>>>([]);
  useEffect(() => {
    datasource.getProjects().then((projects) => setProjects(projects));
  }, [datasource]);

  const projectsWithTemplateVariables = useMemo(
    () => [
      projects,
      {
        label: 'Template Variables',
        options: templateVariableOptions,
      },
      ...projects,
    ],
    [projects, templateVariableOptions]
  );
```

Once that is done, it should be fine to remove the cache completely.

iwysiu committed:
Remove cache and Erik changes

iwysiu commented on pkg/tsdb/cloudmonitoring/resource_handler.go:
```diff
@@ -1,94 +1,296 @@
 	"github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter"
 )
 
+var nameExp = regexp.MustCompile(`([^\/]*)\/*$`)
```
done

iwysiu commented on pkg/tsdb/cloudmonitoring/resource_handler.go:
```diff
@@ -1,94 +1,297 @@
+	mux.HandleFunc("/gceDefaultProject", getGCEDefaultProject)
+
+	mux.HandleFunc("/metricDescriptors/", s.resourceHandler(cloudMonitor, processMetricDescriptors))
+	mux.HandleFunc("/services/", s.resourceHandler(cloudMonitor, processServices))
```
done

iwysiu commented on pkg/tsdb/cloudmonitoring/resource_handler.go:
```diff
@@ -1,94 +1,297 @@
+	mux.HandleFunc("/metricDescriptors/", s.resourceHandler(cloudMonitor, processMetricDescriptors))
+	mux.HandleFunc("/services/", s.resourceHandler(cloudMonitor, processServices))
+	mux.HandleFunc("/slos/", s.resourceHandler(cloudMonitor, processSLOs))
+	mux.HandleFunc("/cloudresourcemanager/", s.resourceHandler(resourceManager, processProjects))
```
done

iwysiu reviewed (commented):
Removed the cache and did a force push to resolve a merge conflict.

andresmgot reviewed (approved):
LGTM!

sunker reviewed (approved):
Great stuff @iwysiu! 

When testing this I did find a problem with annotations but that's not related to your changes so I'll create a separate issue for that.
